File: app/src/main/java/com/test/xyz/demo/presentation/common/di/DaggerApplication.java
package com.test.xyz.demo.presentation.common.di;

import android.app.Application;
import android.content.Context;
import android.support.annotation.VisibleForTesting;

import com.test.xyz.demo.presentation.common.di.component.AppComponent;
import com.test.xyz.demo.presentation.common.di.component.DaggerAppComponent;
import com.test.xyz.demo.presentation.common.di.module.AppModule;

public class DaggerApplication extends Application {
    private static AppComponent appComponent;
    private static DaggerApplication instance;

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
        initAppComponents();
    }

    public static DaggerApplication get(Context context) {
        return (DaggerApplication) context.getApplicationContext();
    }

    public AppComponent getAppComponent() {
        return appComponent;
    }

    public void initAppComponents() {
        appComponent = DaggerAppComponent.builder()
                .appModule(new AppModule(this))
                .build();

    }

    /**
     * Visible only for testing purposes.
     */
    @VisibleForTesting
    public void setTestComponent(AppComponent testingComponent) {
        appComponent = testingComponent;
    }
}

File: app/src/main/kotlin/com/test/xyz/demo/presentation/common/di/DaggerApplication.kt
package com.test.xyz.demo.presentation.common.di

import android.app.Application
import android.content.Context
import com.test.xyz.demo.presentation.common.di.component.AppComponent
import com.test.xyz.demo.presentation.common.di.component.DaggerAppComponent
import com.test.xyz.demo.presentation.common.di.module.AppModule

class DaggerApplication : Application() {

    override fun onCreate() {
        super.onCreate()
        instance = this
        initAppComponents()
    }

    fun getAppComponent(): AppComponent {
        return appComponent
    }

    fun initAppComponents() {
        appComponent = DaggerAppComponent.builder()
                .appModule(AppModule(this))
                .build()

    }

    companion object {
        private lateinit var appComponent: AppComponent
        private lateinit var instance: DaggerApplication

        operator fun get(context: Context): DaggerApplication {
            return context.applicationContext as DaggerApplication
        }
    }
}
